src/main/flight/position.c

```diff
@@ -22,205 +22,202 @@
-    const float altitudeDerivativeGain = pt2FilterGain(altitudeDerivativeCutoffHz, sampleTimeS);
-    pt2FilterInit(&altitudeDerivativeLpf, altitudeDerivativeGain);
-}
+#ifdef USE_GPS
+    // initialize GPS upsampling filter
+    const float gpsRateHz = 1.0f / getGpsDataIntervalSeconds();
+    cutoffHz = gpsRateHz / 4.0f;    // -3dB cutoff at nyquist/2
+    gain = pt2FilterGain(cutoffHz, sampleTimeS);
+    pt2FilterInit(&gpsUpsampleLpf, gain);
+#endif
 
-typedef enum {
-    DEFAULT = 0,
-    BARO_ONLY,
-    GPS_ONLY
-} altitudeSource_e;
+#ifdef USE_BARO
+    // initialize baro upsampling filter
+    const float baroRateHz = 1.0f / ((baro.dev.up_delay + 1000 * (BARO_STATE_COUNT - 1)) * 1e-6f);  // up_delay + 5x 1000us baro state delay
+    cutoffHz = baroRateHz / 4.0f;   // -3dB cutoff at nyquist/2
+    gain = pt2FilterGain(cutoffHz, sampleTimeS);
+    pt2FilterInit(&baroUpsampleLpf, gain);
+#endif
 
-PG_REGISTER_WITH_RESET_TEMPLATE(positionConfig_t, positionConfig, PG_POSITION, 4);
+#if defined(USE_GPS) && defined(USE_BARO)
+    // initialize complementary filter for fusing GPS and barometer
+    cutoffHz = gpsRateHz / 2.0f * positionConfig()->altitude_fuse_ratio / 100.0f;
+    gain = pt1FilterGain(cutoffHz, sampleTimeS);
+    pt1FilterInit(&baroFuseLpf, gain);
+    pt1FilterInit(&gpsFuseLpf, gain);
+#endif
 
-PG_RESET_TEMPLATE(positionConfig_t, positionConfig,
-    .altitude_source = DEFAULT,
-    .altitude_prefer_baro = 100, // percentage 'trust' of baro data
-    .altitude_lpf = 300,
-    .altitude_d_lpf = 100,
-);
+#ifdef USE_VARIO
+    cutoffHz = positionConfig()->altitude_vario_lpf / 100.0f;
+    gain = pt2FilterGain(cutoffHz, sampleTimeS);
+    pt2FilterInit(&varioLpf, gain);
+#endif
+}
 
-#if defined(USE_BARO) || defined(USE_GPS)
-void calculateEstimatedAltitude(void)
+void positionUpdate(void)
 {
-    static bool wasArmed = false;
-    static bool useZeroedGpsAltitude = false; // whether a zero for the GPS altitude value exists
-    static float gpsAltCm = 0.0f; // will hold last value on transient loss of 3D fix
-    static float gpsAltOffsetCm = 0.0f;
-    static float baroAltOffsetCm = 0.0f;
-    static float newBaroAltOffsetCm = 0.0f;
-
-    float baroAltCm = 0.0f;
-    float gpsTrust = 0.3f; // if no pDOP value, use 0.3, intended range 0-1;
-    bool haveBaroAlt = false; // true if baro exists and has been calibrated on power up
-    bool haveGpsAlt = false; // true if GPS is connected and while it has a 3D fix, set each run to false
-
-    // *** Get sensor data
+    // ***  UPSAMPLE SENSOR DATA  ***
 #ifdef USE_BARO
     if (sensors(SENSOR_BARO)) {
-        baroAltCm = getBaroAltitude();
-        haveBaroAlt = true; // false only if there is no sensor on the board, or it has failed
+        isBaroAvailable = isBaroReady(); // false only if no baro is on board, or if baro init has failed
+        baroAltCm = pt2FilterApply(&baroUpsampleLpf, baroGetAltitudeCm());
```

Comment:
```suggestion
        baroAltCm = isBaroAvailable ? pt2FilterApply(&baroUpsampleLpf, baroGetAltitudeCm()) : <some default of 0.0f>;
```

Reply:
`baroGetAltitudeCm()` already returns 0 if the baro is not ready

Reply:
Yes. But why? isBaroAvailable is false, so technically baroGetAltitudeCm() can return anything (this code is dependent on implementation detail)